interfaces/system_key.go

```diff
@@ -80,39 +119,86 @@ func generateSystemKey() *systemKey {
 		logger.Noticef("cannot determine root filesystem on overlay in generateSystemKey: %v", err)
 	}
 
-	// Add the current Core path, we need this because we call helpers
-	// like snap-confine from core that will need an updated profile
-	// if it changes
-	//
-	// FIXME: what about core18? the snapd snap?
-	sk.Core, _ = os.Readlink(filepath.Join(dirs.SnapMountDir, "core/current"))
-
 	// Add seccomp-features
 	sk.SecCompActions = release.SecCompActions()
 
-	return &sk
+	return sk
 }
 
-// SystemKey outputs a string that identifies what security profiles
-// environment this snapd is using. Security profiles that were generated
-// with a different Systemkey should be re-generated.
-func SystemKey() string {
-	sk := generateSystemKey()
+// WriteSystemKey will write the current system-key to disk
+func WriteSystemKey() error {
+	sks, err := json.Marshal(generateSystemKey())
+	if err != nil {
+		panic(err)
+	}
+	return osutil.AtomicWriteFile(dirs.SnapSystemKeyFile, sks, 0644, 0)
+}
 
-	// special case: unknown build-ids always trigger a rebuild
-	if sk.BuildID == "" {
-		return ""
+// SystemKeyMismatch checks if the running binary expects a different
+// system-key than what is on disk.
+//
+// This is used in two places:
+// - snap run: when there is a mismatch it will wait for snapd
+//             to re-generate the security profiles
+// - snapd: on startup it checks if the system-key has changed and
+//          if so re-generate the security profiles
+//
+// This ensures that "snap run" and "snapd" have a consistent set
+// of security profiles. Without it we may have the following
+// scenario:
+// 1. snapd gets refreshed and snaps need updated security profiles
+//    to work (e.g. because snap-exec needs a new permission)
+// 2. The system reboots to start the new snapd. At this point
+//    the old security profiles are on disk (because the new
+//    snapd did not run yet)
+// 3. Snaps that run as daemon get started during boot by systemd
+//    (e.g. network-manager). This may happen before snapd had a
+//    chance to refresh the security profiles.
+// 4. Because the security profiles are for the old version of
+//    the snaps that run before snapd fail to start. For e.g.
+//    network-manager this is of course catastrophic.
+// To prevent this, in step(4) we have this wait-for-snapd
+// step to ensure the expected profiles are on disk.
+func SystemKeyMismatch() (bool, error) {
+	mySystemKey := generateSystemKey()
+
+	raw, err := ioutil.ReadFile(dirs.SnapSystemKeyFile)
+	if err != nil && os.IsNotExist(err) {
+		return false, ErrSystemKeyMissing
 	}
-	sks, err := yaml.Marshal(sk)
 	if err != nil {
-		panic(err)
+		return false, err
+	}
+	var diskSystemKey systemKey
+	if err := json.Unmarshal(raw, &diskSystemKey); err != nil {
+		return false, err
```

Comment:
because previous versions of snapd were writing yaml, will we not hit this which means we will fail to wait for the new profiles on 2.31->2.32?

Reply:
Yes, this is rather critical indeed. I think because we now have a second line of defense by talking to snapd we can simply on error try to reach snapd. This should be ok in the yaml -> json transition as it means that by the time we reach the new snapd the transition will have happend or at least we don't block because we can reach snapd.